colcon and ROS2 support

ipa-mdl opened the pull request:
**target branch:** colcon

This PR adresses #358 and #333 
Most things should work as expected, but might require some config changes.

For the time being I had to remove the support for `CATKIN_CONFIG` because it would only apply for catkin-tools, but not colcon.
Instead, I would like to implement more specific options (#360)

ToDos:

- [x] final update of documentation
- [x] legacy varable detection, warning or error 
- [x] migration hints
- [x] make `industrial_ci` a hybrid (catkin/ament) package
- [ ] clean-up of tests
- [x] ROS2 support in abi check?

ipa-mdl commented:
(PR drafts do not work with Travis as expected: https://travis-ci.community/t/draft-pull-requests-not-being-built/2434)

ipa-mdl commented:
@130s, @miguelprada: do you have some time to review this?
I'd like to roll this out for ROS2 dashing.

ipa-mdl commented:
Addressed #360

miguelprada commented on doc/index.rst:
```diff
@@ -161,70 +164,152 @@ Required environment variables:
+
+For backwards compatibility, ``UPSTREAM_WORKSPACE`` can be set to ``debian`` and ``file`` as well, but not in combination with the other options and with a deprecation warning.
+In case of ``file``, it will be replaced by ``$ROSINSTALL_FILENAME`` or ``$ROSINSTALL_FILENAME.$ROS_DISTRO``, if the latter exists.
+**In ``file`` mode the target repository will not get removed automatically anymore and therefore might get built twice!**
```
I don't understand what you mean with this. Would you mind elaborating a bit?

ipa-mdl replied:
https://github.com/ros-industrial/industrial_ci/blob/master/industrial_ci/src/tests/source_tests.sh#L128 was removed
Without this feature the legacy version would end up in having two packages with the same name -> error.

In the new version these two packages would end up in different workspaces -> no error, but packages in target repositories get build twice.

miguelprada replied:
I wasn't familiar with that feature, since we are used to include `.rosinstall` files per-repository which, obviously, do not include the target repository. I guess that makes sense in configurations that point to a common `rosinstall` for many repositories.

I get the point of the comment now, thanks!

miguelprada commented on doc/index.rst:
```diff
@@ -494,36 +557,40 @@ Run pre/post-process custom commands
-If what you want to customize is within the `CI process <#what-are-checked>`_, you can specify the script(s) in `BEFORE_SCRIPT` and/or `AFTER_SCRIPT` variables. For example::
+If what you want to customize is within the `CI process <#what-are-checked>`__, you can specify the script(s) in `BEFORE_*` and/or `AFTER_*` variables.
+The variables can be set for all functions, using the upper-case name, e.g. to run a script before `install_target_dependencies` you can specify `BEFORE_INSTALL_TARGET_DEPENDENCIES` or `AFTER_INSTALL_TARGET_DEPENDENCIES` to be run afterrwards.
+`INIT_SCRIPT` will be run before any test, `AFTER_SCRIPT` can be used to specify as script to be run after all tests.
```
...to specify _a_ script...

ipa-mdl commented on doc/migration_guide.md:
```diff
@@ -0,0 +1,201 @@
+
+The workspace layout has changed.
+The target workspace is now located in `~/target_ws`.
+Furthermore, the upstream workspace packages are now located in `~/upstream_ws`.
```
There is no downstream set-up in  the legacy version ;)
I can link the full workspace documentation here.

miguelprada replied:
I see your point. Feel free to add the link to the full doc or resolve the conversation as-is.

ipa-mdl commented on doc/index.rst:
```diff
@@ -161,70 +164,152 @@ Required environment variables:
 * **CLANG_FORMAT_VERSION** (default: not set): Version of clang-format to install and use (relates to both the apt package name as well as the executable), e.g., ``CLANG_FORMAT_VERSION=3.8``.
 * **DEBUG_BASH** (default: not set): If set with any value (e.g. `true`), all executed commands that are not printed by default to reduce print space will be printed.
-* **DOCKER_BASE_IMAGE** (default: $OS_NAME:$OS_CODE_NAME): Base image used for building the CI image. Could be used to pre-bundle dependecies or to run tests for different architectures. See `this PR <https://github.com/ros-industrial/industrial_ci/pull/174>`_ for more info.
+* **DOCKER_BASE_IMAGE** (default: $OS_NAME:$OS_CODE_NAME): Base image used for building the CI image. Could be used to pre-bundle dependecies or to run tests for different architectures. See `this PR <https://github.com/ros-industrial/industrial_ci/pull/174>`__ for more info.
```
The value is still set this way, but it won't get used anymore if `ros:$ROS_DISTRO-ros-core` is available

miguelprada replied:
Understood

ipa-mdl commented:
Thanks @miguelprada!
I will update the migration guide in the next days.

> I failed to find a way to automatically determine which workspace my script should process, and ended up duplicating our custom CI setup script with the paths of upstream and target workspaces hardcoded.

```sh
AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  "$upstream_ws"'
AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  "$target_ws"'
```

>  Is there currently any way to find the relevant workspace to act upon on each of these hooks?

We could track a variable for this purpose, but this would only work for some hooks.
As an alternative  a `current_ws` function could be implemented (similar to `rosenv`).

> Default ros:$ROS_DISTRO-ros-core images do not contain an SSH client, [...]  but there must be a better way..

I refactored the code a little bit and implemented a common init step.
You could just specify `ADDITIONAL_DEBS=openssh-client` in all tests.
It migth make sense to install the client automatically if the `.ssh` folder exists (in the container).

> Clang format test errors due to the LLVM format test code have re-appeared (see this job). I've had a look to see whether I could figure out why, but I failed miserably...

I will have look.

ipa-mdl commented:
> My last comment is with regard to eliminating the job control-related configuration that exists in the legacy version. I guess this may be related to introducing a new builder and the fact that variables existing in legacy may not directly map to it.

In the first place I removed everything that was ROS1/catkin-specific.

> However, we currently rely on these to avoid exhausting the limited resources the machine running our pipelines has at its disposal. I will need to think a bit more about this, and try to figure out whether we can achieve the same result by other means, but have you got any ideas yourself about how this feature could be introduced back?

I was hoping that the builders would auto-scale..
We can reintroduce this setting somehow, e.g. `BUILDER_JOBS=2` `BUILDER_TEST_JOBS=1`.
However, I don't know if we could map it properly.

Another option is to play with the Docker limits.

miguelprada commented:
> ```sh
> AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  "$upstream_ws"'
> AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  "$target_ws"'
> ```

Will give this a go.

> I was hoping that the builders would auto-scale..

They probably do for a given job based on the number of available CPU cores, but we were trying to run more than one job concurrently, and I'm not sure builders will be able to properly scale in this context.

> We can reintroduce this setting somehow, e.g. BUILDER_JOBS=2 BUILDER_TEST_JOBS=1.
> However, I don't know if we could map it properly.
> 
> Another option is to play with the Docker limits.

Will think about this, discuss it with my colleagues and try to come up with a solution that works for us.

ipa-mdl commented:
> Clang format test errors due to the LLVM format test code have re-appeared (see this job). I've had a look to see whether I could figure out why, but I failed miserably...

Please test again.

miguelprada commented:
> > ```sh
> > AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  "$upstream_ws"'
> > AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  "$target_ws"'
> > ```
> 
> Will give this a go.

I'm not able to make this work. Since `upstream_ws` and `target_ws` are declared `local`, I guess they are not accessible at the point of processing the hook. Is this correct?

I can, alternatively, hardcode the paths at the level of the CI config, which is a bit better than duplicating the scripts.

```sh
AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  /root/upstream_ws'
AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  /root/target_ws'
```

I can live with this, but if you can think of a better way I'll be glad to test it out.

> > Clang format test errors due to the LLVM format test code have re-appeared (see this job). I've had a look to see whether I could figure out why, but I failed miserably...
> 
> Please test again.

The clang-format check now [passed](https://travis-ci.org/miguelprada/ur_modern_driver/jobs/531972924)! :+1:

ipa-mdl commented:
> I'm not able to make this work. Since upstream_ws and target_ws are declared local, I guess they are not accessible at the point of processing the hook. Is this correct?

Oh, that's right.
We could export them though.
Futhermore, I'd like to make at least target_ws renameable.

> but if you can think of a better way I'll be glad to test it out

How do you like https://github.com/ipa-mdl/industrial_ci/commit/4a95c95233d548683ef68df4fcca69321a0f235e?

miguelprada commented:
It works for me, but I have to use double `$$` to prevent [GitLab substituting the variable](https://gitlab.com/gitlab-org/gitlab-ce/issues/27436) way too early, i.e.:

```sh
AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  $$upstream_ws'
AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  $$target_ws'
```

ipa-mdl commented:
```
AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  $$current_ws'
AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  $$current_ws'
```
works now as well.

miguelprada commented:
> ```
> AFTER_SETUP_UPSTREAM_WORKSPACE='./hook.sh  $$current_ws'
> AFTER_SETUP_TARGET_WORKSPACE='./hook.sh  $$current_ws'
> ```
> 
> works now as well.

Doesn't work for me: `current_ws` returns `/root/target_ws` even when called in `AFTER_SETUP_UPSTREAM_WORKSPACE`. I think it's due to a typo in [this line](https://github.com/ipa-mdl/industrial_ci/blob/c13f9533510f902de9aa4862bbfb4fac375d55ef/industrial_ci/src/tests/source_tests.sh#L51). That should be:

```sh
ici_with_ws "$upstream_ws" ici_build_workspace "upstream" "$extend" "$upstream_ws"
```

instead of

```sh
ici_with_ws "$target_ws" ici_build_workspace "upstream" "$extend" "$upstream_ws"
```

ipa-mdl commented:
I pushedt he fix. In general I won't rebase anymore for now to simplify the review.

130s commented:
> @130s, @miguelprada: do you have some time to review this?

I try not to review while a PR is WIP (had some bad experience before). If this is not WIP please remove WIP.

ipa-mdl commented:
> If this is not WIP please remove WIP.

The commit structure is still WIP, so I cannot remove this for now.
In addition I could not test all cases, so I depend on your test config.
I will update the code as needed to fix your tests.
However, #369 already contains commits that are not WIP anymore.

ipa-mdl commented:
This is now rebased and ready for merge.

Things to review:
- inspect test jobs outputs
- spelling in documentation / migration guide
- completeness of migration guide
- test industrial_ci locally with ROS2.
- source code review if possible

I'd like to merge and announce this before the release of ROS2 dashing, as an open beta test.

Things to be taken care of later:

- Use default dashing image
- Switch dashing test back to `control_msgs` or a different package
- test kinetic after https://github.com/ros/catkin/issues/999
- Update tests for better coverage and perhaps speed-up

ipa-mdl commented:
Thanks for your review!

> One thing I couldn't tell; I'm afraid some existing usecases will fail because associated variables are getting rid of in this PR.

Yes, it might break some set-ups.
But for most users there is an easy way to migrate.

> I see this PR is not merged into a branch mainly advertised (yet), so I assume merging this PR won't break any existing usecases.

This has been discussed [a lot](https://github.com/ros-industrial/industrial_ci/issues/358).
The current plan is to merge this into `master`. I have updated the PR. 

> But when we sync these changes into a main branch, that would happen. What's our opnion for that?

The `master` version should be opt-in for a while..
Actually, I wanted to start this beta phase last month, but review is still pending ;)

ipa-mdl commented on doc/migration_guide.md:
```diff
@@ -0,0 +1,269 @@
+The new version comes with a number of improvements:
+
+* Support for catkin_make, catkin_make_isolated and colcon
+* Support for ROS2, and vcstool (rosinstall or repos files)
```
I think it is fair to mention it in the migration guide.
We can link and refer to the "vsc" command to make it more relatable.

ipa-mdl replied:
done

ipa-mdl commented on doc/migration_guide.md:
```diff
@@ -0,0 +1,269 @@
+
+The new version comes with a number of improvements:
+
+* Support for catkin_make, catkin_make_isolated and colcon
```
> This sentence makes me wonder if catkin_tools support is dropped (I don't think that's the case).

Of course not. But obviously, it won't work for ROS2.

> catkin_tools is still supported (and default).

Actually, `colcon` might become the default at some point.

ipa-mdl replied:
just prefixed it with "Added"

ipa-mdl commented on industrial_ci/src/tests/source_tests.sh:
```diff
@@ -22,259 +22,60 @@
-    ici_time_end  # catkin_run_tests
-fi
 
-if [ "$NOT_TEST_INSTALL" != "true" ]; then
```
There is no ROS2 equivalent.
And I doubt that anyone is using this feature (see https://github.com/ros-industrial/industrial_ci/issues/252 as well).

ipa-mdl commented on doc/index.rst:
```diff
@@ -189,30 +187,124 @@ Note that some of these currently tied only to a single option, but we still lea
+  For convenience ``@`` (pip-style) can be used as a separator as well.
+
+* URL (=starts with http or https) of a ``.repos`` or ``*.rosinstall`` file
+* relative path of a ``.repos`` or ``*.rosinstall`` file
```
Good point, I will add the wildcard

ipa-mdl replied:
added the wildcards

ipa-mdl commented on doc/index.rst:
```diff
@@ -189,30 +187,124 @@ Note that some of these currently tied only to a single option, but we still lea
+  UPSTREAM_WORKSPACE='github:ros-controls/ros_control#melodic-devel'
+
+
+To depend on a different GitHub repository, e.g. ros_control, but only a subset of it:
```
>  could be interpreted as either ex/include.

But this would be wrong. It actually removes the folder recursively.

130s replied:
```
UPSTREAM_WORKSPACE='github:ros-controls/ros_control#melodic-devel -rqt_controller_manager'
```
What I meant is that users can get confused with this example whether only `rqt_controller_manager` will remain or disappear. That said, I think this is minor.

ipa-mdl commented on doc/index.rst:
```diff
@@ -189,30 +187,124 @@ Note that some of these currently tied only to a single option, but we still lea
+
+Each workspace can be composed as a sequence of the folling items:
+
+* URL of a source repository with the pattern ``<scheme>:<resource>#<version>``, e.g. ``github:ros-industrial/industrial_ci#master``.
```
It fails.
```
ici_error "could not parse URL '$url'"
```

ipa-mdl replied:
I added some notes regarding the version

ipa-mdl commented on doc/index.rst:
```diff
@@ -502,22 +572,26 @@ You may want to add custom steps prior/subsequent to the setup defined in `indus
 
-If what you want to customize is within the `CI process <#what-are-checked>`__, you can specify the script(s) in `BEFORE_SCRIPT` and/or `AFTER_SCRIPT` variables. For example::
+If what you want to customize is within the `CI process <#what-are-checked>`__, you can specify the script(s) in `BEFORE_*` and/or `AFTER_*` variables.
+The variables can be set for all functions, using the upper-case name, e.g. to run a script before `install_target_dependencies` you can specify `BEFORE_INSTALL_TARGET_DEPENDENCIES` or `AFTER_INSTALL_TARGET_DEPENDENCIES` to be run afterrwards.
```
The get printed in the log.
But we should document this somehow (later).

ipa-mdl commented on doc/migration_guide.md:
```diff
@@ -0,0 +1,269 @@
+
+Furthermore, the hook will get run without a ROS environment (`setup.bash`).
+If you need this environment, you can use the `rosenv` helper.
+It will crawl all workspaces from `$downstream_ws` to `/opt/ros/$ROS_DISTRO` for the a `setup.bash` file and will source the first one.
```
More likely:

OLD: not defined sourcing of the workspace
NEW: explicit sourcing of the currently active workspace.

PS: We might want `rosenv` use `current_ws`

ipa-mdl replied:
I just removed this, it might confuse more than it helps.

> We might want rosenv use current_ws

done

ipa-mdl commented:
rebased on top of #378

ipa-mdl commented on doc/index.rst:
```diff
@@ -189,30 +187,124 @@ Note that some of these currently tied only to a single option, but we still lea
+Workspace definition
+++++++++++++++++++++
+
+Each workspace can be composed as a sequence of the folling items:
```
fixed

ipa-mdl commented:
I just found out (https://github.com/ros-controls/ros2_controllers/pull/6) that rosdep does not work properly with chained colcon workspaces

miguelprada commented:
> I just found out (ros-controls/ros2_controllers#6) that rosdep does not work properly with chained colcon workspaces

I'm not sure I got it right, but it seems to be that the `-i` flag is not properly identifying the pakcages in the upstream workspace, isn't it?

If that's the case, it seems that you may be able to get around it by explicitly specifying the upstream source-spaces using `--from-paths`, e.g. `rosdep install -iy --from-pahts /path/to/target_ws/src /path/to/upstream_ws/src`.

ipa-mdl commented:
> I'm not sure I got it right, but it seems to be that the -i flag is not properly identifying the pakcages in the upstream workspace, isn't it?

Kind of. If I am not mistakenn, rosdep is not able to find packages in the upstream workspace, because they are not exposed via `ROS_PACKAGE_PATH`.

> it seems that you may be able to get around it by explicitly specifying the upstream source-spaces using --from-paths, e.g. rosdep install -iy --from-pahts /path/to/target_ws/src /path/to/upstream_ws/src.

Something like this. I did not find a nice way to pass these things down..

130s commented:
https://github.com/ros-controls/ros2_controllers/pull/6 now passed CI. Does that mean rosdep problem https://github.com/ros-industrial/industrial_ci/pull/361#issuecomment-507721062 is fixed?

If there are new commits since the last review I'm happy to take a look. My last only concern is gone https://github.com/ros-industrial/industrial_ci/issues/358#issuecomment-508958414

ipa-mdl commented:
I will rebase this on top of #393 and include the fixup.

> ros-controls/ros2_controllers#6 now passed CI. Does that mean rosdep problem #361 (comment) is fixed?

Yes, more or less.
https://github.com/ros-industrial/industrial_ci/pull/361/commits/ab33c5d75489038dc5748e0d949666fe19a892d8 is fixing this, but I would consider it a hack.

ipa-mdl commented:
According to https://github.com/ros-infrastructure/rosdep/issues/660, `AMENT_PREFIX_PATH` should be the way to go for now.

AAlon commented:
Hey @ipa-mdl, I started using this for prerelease testing on dashing and it seemed to work fine yesterday ([successful build](https://travis-ci.org/AAlon/ROS2Prerelease/jobs/565781955/config)), but then started failing with exceptions from the buildfarm scripts complaining about missing packages:
`KeyError: "The cache has no package named 'ros-dashing-rclcpp'"`
See [build](https://travis-ci.org/AAlon/ROS2Prerelease/jobs/566255915). Any idea what might be wrong?

ipa-mdl commented:
@AAlon: I could not reproduce this error and your build seem to work again..

@miguelprada, @130s: Ready for merge?

ipa-mdl commented:
@miguelprada, @130s: If you do not object, I will merge this by the end of this week.

130s commented:
+1 for merge (although I haven't reviewed after my previous approval). Sorry I've been aware of this but re-reviewing this would probably take good amount of time, and I doubt I can do that in next few weeks. The target branch is not the default so I think less impact to the users. And once merged in I hope I can contribute by using it as we're in need of some new features added.

ipa-mdl commented:
I just rebased the commits a little bit to reflect the latest changes.

@130s: Thanks for your response! I will go ahead tomorrow :)

ipa-mdl commented:
I will keep my `colcon-rebase` branch for a while
